sockeye/inference.py

```diff
@@ -1089,22 +1108,55 @@ def _beam_search(self,
             attentions[:, t, :] = attention_scores
             lengths += mx.nd.cast(1 - mx.nd.expand_dims(finished, axis=1), dtype='float32')
 
-            # (6) determine which hypotheses in the beam are now finished
+
+            # (6) save beam history
+            # TODO hussed: check speed, if slow,  condition this block on some CLI flag so it doesn't run in prod
+            unnormalized_scores = scores_accumulated * self.length_penalty(lengths -1)
+            for sent in range(self.batch_size):
+                rows = slice(sent * self.beam_size, (sent + 1) * self.beam_size)
+
+                best_word_indices_sent = best_word_indices[rows].asnumpy().tolist()
+                # avoid adding columns for finished sentences
+                if any(x for x in best_word_indices_sent if x != C.PAD_ID):
+                    beam_histories[sent]["predicted_ids"].append(best_word_indices_sent)
+                    beam_histories[sent]["predicted_tokens"].append([self.vocab_target_inv[x] for x in
+                                                                best_word_indices_sent])
+                    # for later sentences in the matrix, shift from e.g. [5, 6, 7, 8, 6] to [0, 1, 3, 4, 1]
+                    shifted_parents = best_hyp_indices[rows] - (sent * self.beam_size)
+                    beam_histories[sent]["parent_ids"].append(shifted_parents.asnumpy().tolist())
+
+                    beam_histories[sent]["scores"].append(unnormalized_scores[rows].asnumpy().flatten().tolist())
+                    beam_histories[sent]["normalized_scores"].append(scores_accumulated[rows].asnumpy().flatten().tolist())
+
+
+            # (7) determine which hypotheses in the beam are now finished
             finished = ((best_word_indices == C.PAD_ID) + (best_word_indices == self.vocab_target[C.EOS_SYMBOL]))
             if mx.nd.sum(finished).asscalar() == self.batch_size * self.beam_size:  # all finished
                 break
 
-            # (7) update models' state with winning hypotheses (ascending)
+            # (8) update models' state with winning hypotheses (ascending)
             for ms in model_states:
                 ms.sort_state(best_hyp_indices)
 
-        return sequences, attentions, scores_accumulated, lengths
+        # count the number of steps in each beam
+        for history in beam_histories:
+            history["number_steps"] = len(history["predicted_tokens"])
```

Comment:
isn't that exactly the `t` variable from beam search?

Reply:
Each history corresponds to a sentence. In the batching case, this code ensures that `"number_steps"` is right for each sentence within a batch. `t` would just give the right size for the longest sentence in the batch.

Reply:
So you want the length without padding tokens? Then you can use the `lengths` array in the function maybe.